src/components/TradeWidget/TokenRow.tsx

```diff
@@ -113,17 +135,55 @@ function displayBalance(balance: TokenBalanceDetails | undefined | null, key: st
   return formatAmount(balance[key], balance.decimals) || '0'
 }
 
+function getMax(balance: TokenBalanceDetails | null, token: TokenDetails): string {
+  return formatAmount((balance || {}).exchangeBalance, token.decimals, 4, false) || '0'
+}
+
+// TODO: move into a validators file?
+function validatePositive(value: string): true | string {
+  return Number(value) > 0 || 'Invalid amount'
+}
+
 interface Props {
   token: TokenDetails
   tokens: TokenDetails[]
   balance: TokenBalanceDetails
   selectLabel: string
   onSelectChange: (selected: TokenDetails) => void
+  inputId: string
+  validateMaxAmount?: true
 }
 
-const TokenRow: React.FC<Props> = ({ token, tokens, selectLabel, onSelectChange, balance }) => {
+const TokenRow: React.FC<Props> = ({
+  token,
+  tokens,
+  selectLabel,
+  onSelectChange,
+  balance,
+  inputId,
+  validateMaxAmount,
+}) => {
   const options = useMemo(() => tokens.map(token => ({ token, value: token.symbol, label: token.name })), [tokens])
 
+  const { register, errors, setValue, watch } = useFormContext()
+  const error = errors[inputId]
+
+  const max = Number(getMax(balance, token))
+  const inputValue = Number(watch(inputId)) || 0
+  const overMax = validateMaxAmount && inputValue > max ? inputValue - max : 0
```

Comment:
As `overMax` error isn't from the hook-form, I don't think it will prevent form submission
If I'm right we need either to especially check for it in `submitForm`, or make it independent from `useFormContext` instance.

I have this suggestion:
```ts
const TokenRow =(...) => {
   const max = getMax(balance, token)

  return <Wrapper>
      <TokenImgWrapper alt={token.name} src={token.image} />
      <SelectBox>
        ...
      </SelectBox>
      <InputBox>
        <InputComp key={token.name + max} max={max} balance={balance}/>
...
}

const InputComp = ({max, ...}) => {
  return       <InputBox>
        <input type="text" placeholder="0" />
        <input
          className={className}
          placeholder="0"
          name={inputId}
          type="text"
          required
          ref={register({
            validate: {
              positive: (value: string): true | string => value < max || 'Not enough balance'
            },
          })}
        />
        <WalletDetail>
            ...
        </WalletDetail>
}
```

Including `key` in props will remount component on key change, so register will always get an up-to-date validation rule and there'll be no need for out-of-form error checks.

Or that's the idea, not sure if it'll work.

Reply:
Same can be done with `register/unregister` in `useEffect(..., [max])` but it's ugly.